File: range_sum_query_2d_immutable/range_sum_query_2d_immutable.py
#!/usr/bin/env python3
"""range_sum_query_2d_immutable.py"""
from typing import List
from timeit import timeit


class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        """Initializes the object with the
        integer matrix matrix and pre-computes
        all matrix subset sums into
        self.sub_matrices."""
        self.matrix = matrix
        rows, cols = len(matrix) + 1, len(matrix[0]) + 1
        self.sub_matrices = [[0] * cols for _ in range(rows)]
        for row in range(1, rows):
            for col in range(1, cols):
                self.sub_matrices[row][col] = (
                    self.sub_matrices[row - 1][col] +
                    self.sub_matrices[row][col - 1] -
                    self.sub_matrices[row - 1][col - 1] +
                    matrix[row - 1][col - 1])

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        """Returns the sum of the elements of matrix inside the rectangle
        defined by its upper left corner (row1, col1) and lower right corner
        (row2, col2)."""
        return (self.sub_matrices[row2 + 1][col2 + 1] -
                self.sub_matrices[row2 + 1][col1] -
                self.sub_matrices[row1][col2 + 1] +
                self.sub_matrices[row1][col1])

    def sumRegion2(self, row1: int, col1: int, row2: int, col2: int) -> int:
        """Returns the sum of the elements of matrix inside the rectangle
        defined by its upper left corner (row1, col1) and lower right corner
        (row2, col2)."""
        sub_matrix = 0
        for i in self.matrix[row1:(row2 + 1)]:
            sub_matrix += sum(i[col1:(col2 + 1)])
        return sub_matrix


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)

TEST_MATRIX = [[3, 0, 1, 4, 2], [5, 6, 3, 2, 1], [1, 2, 0, 1, 5],
               [4, 1, 0, 1, 7], [1, 0, 3, 0, 5]]
numMatrix = NumMatrix(TEST_MATRIX)

assert numMatrix.sumRegion(2, 1, 4, 3) == 8
assert numMatrix.sumRegion(1, 1, 2, 2) == 11
assert numMatrix.sumRegion(1, 2, 2, 4) == 12

assert numMatrix.sumRegion2(2, 1, 4, 3) == 8
assert numMatrix.sumRegion2(1, 1, 2, 2) == 11
assert numMatrix.sumRegion2(1, 2, 2, 4) == 12

X = timeit(lambda: numMatrix.sumRegion(1, 2, 2, 4), number=1000000)
Y = timeit(lambda: numMatrix.sumRegion2(1, 2, 2, 4), number=1000000)
print(X)
print(Y)

File: range_sum_query_2d_immutable/range_sum_query_2d_immutable.c
#include <stdio.h>
#include <stdlib.h>

// gcc range_sum_query_2d_immutable.c -O1 -fsanitize=address -fno-omit-frame-pointer -std=gnu99; ./a.out


typedef struct {
    int **matrix;
    int matrixSize;
} NumMatrix;

NumMatrix* numMatrixCreate(int** matrix, int matrixSize, int* matrixColSize) {
    /* Pre-computes all matrix subset sums into subMatrices of
     * type NumMatrix. */
    NumMatrix *subMatrices = malloc((matrixSize) * sizeof(NumMatrix));
    subMatrices->matrix = malloc((matrixSize + 1) * sizeof(int*));
    subMatrices->matrixSize = (matrixSize + 1);
    int i, j;
    for(i = 0; i <= matrixSize; i++){
        subMatrices->matrix[i] = calloc((*matrixColSize + 1), sizeof(int));
    }
    for(i = 1; i <= matrixSize; i++){
        for(j = 1; j <= *matrixColSize; j++){
            subMatrices->matrix[i][j] = (
                    subMatrices->matrix[i - 1][j] +
                    subMatrices->matrix[i][j - 1] -
                    subMatrices->matrix[i - 1][j - 1] +
                    matrix[i - 1][j - 1]
                    );

        }
    }
    return subMatrices;
}

int numMatrixSumRegion(NumMatrix* obj, int row1, int col1, int row2, int col2) {
    /* Returns the sum of the elements of matrix inside the rectangle
     * defined by its upper left corner (row1, col1) and lower right corner. */
    return (
            obj->matrix[row2 + 1][col2 + 1] -
            obj->matrix[row2 + 1][col1] -
            obj->matrix[row1][col2 + 1] +
            obj->matrix[row1][col1]
           );
}

void numMatrixFree(NumMatrix* obj) {
    /* Free memory allocated for NumMatrix object. */
    int i;
    for(i = 0; i < obj->matrixSize; i++){
        free(obj->matrix[i]);
    }
    free(obj->matrix);
    free(obj);
}

/**
 * Your NumMatrix struct will be instantiated and called as such:
 * NumMatrix* obj = numMatrixCreate(matrix, matrixSize, matrixColSize);
 * int param_1 = numMatrixSumRegion(obj, row1, col1, row2, col2);

 * numMatrixFree(obj);
*/

// TESTS
int main(void){
    int returnVal = 0;
    int result;
    int mS = 5;
    int mCS = 5;
    int a[5] = {3, 0, 1, 4, 2};
    int b[5] = {5, 6, 3, 2, 1};
    int c[5] = {1, 2, 0, 1, 5};
    int d[5] = {4, 1, 0, 1, 7};
    int e[5] = {1, 0, 3, 0, 5};
    int **matrix = malloc(mS * sizeof(int*));
    matrix[0] = a, matrix[1] = b, matrix[2] = c, matrix[3] = d, matrix[4] = e;
    NumMatrix *testMatrix = numMatrixCreate(matrix, mS, &mCS);

    // PRINTS FOR DEV DEBUGGING //
    /*
    int i, j;
    for(i = 0; i <= mS; i++){
        for(j = 0; j <= mCS; j++){
            printf("%d ", testMatrix->matrix[i][j]);
        }
    }
    printf("\n %d\n", numMatrixSumRegion(testMatrix, 2, 1, 4, 3));
    */

    // assert [2, 1, 4, 3] == 8
    result = numMatrixSumRegion(testMatrix, 2, 1, 4, 3);
    if(result != 8){returnVal = 1;}
    printf("\n%d", result);

    //assert [1, 1, 2, 2] == 11
    result = numMatrixSumRegion(testMatrix, 1, 1, 2, 2);
    if(result != 11){returnVal = 1;}
    printf("\n%d", result);

    //assert [1, 2, 2, 4] == 12
    result = numMatrixSumRegion(testMatrix, 1, 2, 2, 4);
    if(result != 12){returnVal = 1;}
    printf("\n%d", result);

    printf("\n");
    numMatrixFree(testMatrix);
    free(matrix);
    return returnVal;
}
